lib/msf/core/payload/adapter/fetch.rb

```diff
@@ -0,0 +1,316 @@
+    super
+    register_options(
+      [
+        Msf::OptEnum.new('FETCH_COMMAND', [true, 'Command to fetch payload', 'CURL', %w{ WGET CURL FTP TNFTP TFTP Certutil Psh_InvokeWebRequest }]),
+        Msf::OptPort.new('FETCH_SRVPORT', [true, 'Local port to use for serving payload', 8080]),
+        Msf::OptAddressLocal.new('FETCH_SRVHOST', [ true, 'Local IP to use for serving payload', "0.0.0.0"]),
+        Msf::OptString.new('FETCH_URIPATH', [ false, 'Local URI to use for serving payload', '']),
+        Msf::OptString.new('FETCH_WRITABLE_DIR', [ true, 'Remote writable dir to store payload; cannot contain spaces.', ''], regex:/^[\S]*$/),
+        Msf::OptString.new('FETCH_FILENAME', [ false, 'Name to use on remote system when storing payload; cannot contain spaces.', Rex::Text.rand_text_alpha(rand(8..12))], regex:/^[\S]*$/)
+      ]
+    )
+    register_advanced_options(
+      [
+        Msf::OptAddress.new('FetchListenerBindAddress', [ false, 'The specific IP address to bind to to serve the payload if different from FETCH_SRVHOST']),
+        Msf::OptPort.new('FetchListenerBindPort', [false, 'The port to bind to if different from FETCH_SRVPORT']),
+        Msf::OptBool.new('FetchHandlerDisable', [true, 'Disable fetch handler', false]),
+        Msf::OptString.new('FetchServerName', [true, 'Fetch Server Name', 'Apache'])
+      ]
+    )
+    @srvexe = ''
+    @myresources = []
+
+  end
+
+  def check_srvhost
+    if Rex::Socket.is_ip_addr?(srvhost) && Rex::Socket.addr_atoi(srvhost) == 0
+      raise ArgumentError, 'You must set FETCH_SRVHOST to a routable IP'
+    end
+  end
+
+  def compatible?(mod)
+    if mod.type == Msf::MODULE_PAYLOAD && (mod.class.const_defined?(:CachedSize) && mod.class::CachedSize != :dynamic) && (mod.class::CachedSize >= 120_000) # echo does not have an unlimited amount of space
+      return false
+    end
+    super
+  end
+
+
+  # If no fetch URL is provided, we generate one based off the underlying payload data
+  # This is because if we use a randomly-generated URI, the URI generated by venom and
+  # Framework will not match.  This way, we can build a payload in venom and a listener
+  # in Framework, and if the underlying payload type/host/port are the same, the URI
+  # will be, too.
+  #
+  def default_srvuri
+    # If we're in framework, payload is in datastore; msfvenom has it in refname
+    payload_name = datastore['payload'] ||= refname
+    decoded_uri = payload_name.dup
+    # there may be no transport, so leave the connection string off if that's the case
+    if module_info['ConnectionType'].upcase == 'REVERSE'
+      decoded_uri << "[#{datastore['LHOST']}]" unless datastore['LHOST'].nil?
+      decoded_uri << datastore['LPORT'].to_s unless datastore['LPORT'].nil?
+    end
+    if module_info['ConnectionType'].upcase == 'BIND'
+      decoded_uri << "[#{datastore['LHOST']}]" unless datastore['LHOST'].nil?
+      decoded_uri << datastore['RPORT'].to_s unless datastore['RPORT'].nil?
+    end
```

Comment:
This change switches to using brackets only when necessary because the host is an IPv6 address and uses a semicolon as a delimiter between the URI and host information.

```suggestion
    netloc = ''
    if module_info['ConnectionType'].upcase == 'REVERSE'
      netloc << datastore['LHOST'] unless datastore['LHOST'].blank?
      unless datastore['LPORT'].blank?
        if Rex::Socket.is_ipv6?(netloc)
          netloc = "[#{netloc}]:#{datastore['LPORT']}"
        else
          netloc = "#{netloc}:#{datastore['LPORT']}"
        end
      end
    elsif module_info['ConnectionType'].upcase == 'BIND'
      netloc << datastore['LHOST'] unless datastore['LHOST'].blank?
      unless datastore['RPORT'].blank?
        if Rex::Socket.is_ipv6?(netloc)
          netloc = "[#{netloc}]:#{datastore['RPORT']}"
        else
          netloc = "#{netloc}:#{datastore['RPORT']}"
        end
      end
    end
    decoded_uri << ";#{netloc}"
```

Reply:
Done